File: captchatools/capmonster.py
from .errors import ErrCodeToException
from . import Harvester
import time
import requests
from typing import Optional

BASEURL = "https://api.capmonster.cloud"



class Capmonster(Harvester):
    '''
    This object will contain the data to interact with capmonster.cloud API
    '''
    def get_balance(self) -> float:
        payload = {"clientKey": self.api_key}
        for _ in range(5):
            try:
                resp = requests.post(BASEURL + "/getBalance", json=payload ,timeout=20).json()
                if resp["errorId"] == 1: # Means there was an error
                    ErrCodeToException("Capmonster", resp["errorCode"])
                return float(resp["balance"])
            except requests.RequestException:
                pass

    def get_token(self, b64_img: Optional[str] = None, user_agent: Optional[str] = None, proxy: Optional[str] = None, proxy_type: Optional[str] = "HTTP"):
        # Get ID
        task_id = self.__get_id(
            b64_img=b64_img,
            user_agent=user_agent,
            proxy=proxy,
            proxy_type=proxy_type
        )
        
        # Get Answer
        return self.__get_answer(task_id)
    
    def __create_payload(self, **kwargs):
        payload = {"clientKey":self.api_key,"task":{}}

        # Add data based on the captcha type
        if self.captcha_type == "image" or self.captcha_type == "normal":
            payload["task"]["type"] = "ImageToTextTask"
            payload["task"]["body"] = kwargs.get("b64_img", "")
        elif self.captcha_type == "v2":
            payload["task"]["type"] = "NoCaptchaTaskProxyless"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
            if self.invisible_captcha:
                payload["task"]["isInvisible"] = self.invisible_captcha
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "NoCaptchaTask"
        elif self.captcha_type == "v3":
            payload["task"]["type"] = "RecaptchaV3TaskProxyless"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey
            payload["task"]["minScore"] = self.min_score
            payload["task"]["pageAction"] = self.action
        elif self.captcha_type == "hcap" or self.captcha_type == "hcaptcha":
            payload["task"]["type"] = "HCaptchaTaskProxyless"
            if kwargs.get("proxy", None) is not None:
                payload["task"]["type"] = "HCaptchaTask"
            payload["task"]["websiteURL"] = self.captcha_url
            payload["task"]["websiteKey"] = self.sitekey


        # Add Global Data
        if self.soft_id is not None:
            payload["softId"] = self.soft_id
        if kwargs.get("proxy", None) is not None:
            splitted = kwargs.get("proxy").split(":")
            payload["task"]["proxyAddress"] = splitted[0]
            try:
                payload["task"]["proxyPort"] = int(splitted[1])
            except Exception:
                payload["task"]["proxyPort"] = splitted[1]
            if len(splitted) >=4:
                payload["task"]["proxyLogin"] = splitted[2]
                payload["task"]["proxyPassword"] = splitted[3]
            payload["task"]["proxyType"] = kwargs.get("proxy_type", "http")
        if kwargs.get("user_agent", None) is not None:
            payload["task"]["userAgent"] = kwargs.get("user_agent")
        return payload
    
    def __get_id(self,**kwargs):
        # Create Payload
        payload = self.__create_payload(**kwargs)
        
        # Get token & return it
        for _ in range(50):
            try:
                resp = requests.post(BASEURL + "/createTask" , json=payload, timeout=20).json()
                if resp["errorId"] != 0: # Means there was an error:
                    ErrCodeToException("Capmonster", resp["errorCode"])
                return resp["taskId"]
            except (requests.RequestException, KeyError):
                pass
    
    def __get_answer(self,task_id:int):
        payload = {"clientKey":self.api_key,"taskId": task_id}
        for _ in range(100):
            try:
                response = requests.post(BASEURL + "/getTaskResult",json=payload,timeout=20,).json()
                if response["errorId"] != 0: # Means there was an error
                    ErrCodeToException("Capmonster", response["errorCode"])
                if response["status"] == "processing":
                    time.sleep(3)
                    continue
                if self.captcha_type == "normal" or self.captcha_type == "image":
                    return response["solution"]["text"]
                else:
                    return response["solution"]["gRecaptchaResponse"]
            except (requests.RequestException, KeyError):
                pass

File: captchatools-go/capmonster.go
package captchatoolsgo

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"
)

/*
   This file will contain the code to interact with capmonster.cloud API
*/

type Capmonster struct {
	config *Config
}

// This struct will be the payload to get the queue ID from capmonster
type capmonsterIDPayload struct {
	ClientKey string `json:"clientKey"`
	SoftID    int    `json:"softId,omitempty"`
	Task      struct {
		WebsiteURL    string      `json:"websiteURL"`
		WebsiteKey    string      `json:"websiteKey"`
		Type          captchaType `json:"type"`
		IsInvisible   bool        `json:"isInvisible,omitempty"`
		MinScore      float32     `json:"minScore,omitempty"`
		PageAction    string      `json:"pageAction,omitempty"`
		Body          string      `json:"body,omitempty"`
		ProxyType     string      `json:"proxyType,omitempty"`
		ProxyAddress  string      `json:"proxyAddress,omitempty"`
		ProxyPort     int         `json:"proxyPort,omitempty"`
		ProxyLogin    string      `json:"proxyLogin,omitempty"`
		ProxyPassword string      `json:"proxyPassword,omitempty"`
		UserAgent     string      `json:"userAgent,omitempty"`
	} `json:"task"`
}
type capmonsterCapAnswerPayload struct {
	ClientKey string `json:"clientKey"`
	TaskID    int    `json:"taskId"`
}

type capmonsterIDResponse struct {
	ErrorID   int    `json:"errorId"`
	ErrorCode string `json:"errorCode"`
	TaskID    int    `json:"taskId"`
}

type capmonsterTokenResponse struct {
	ErrorID   int    `json:"errorId"`
	ErrorCode string `json:"errorCode"`
	Solution  struct {
		Token              string `json:"token"`
		Text               string `json:"text"`
		GRecaptchaResponse string `json:"gRecaptchaResponse"`
		UserAgent          string `json:"userAgent"`
	} `json:"solution"`
	Status string `json:"status"`
}
type capmonsterBalanceResponse struct {
	Balance   float32 `json:"balance"`
	ErrorCode string  `json:"errorCode"`
	ErrorID   int     `json:"errorId"`
}

func (c Capmonster) GetToken(additional ...*AdditionalData) (*CaptchaAnswer, error) {
	return c.getCaptchaAnswer(context.Background(), additional...)
}

func (c Capmonster) GetTokenWithContext(ctx context.Context, additional ...*AdditionalData) (*CaptchaAnswer, error) {
	return c.getCaptchaAnswer(ctx, additional...)
}

func (c Capmonster) GetBalance() (float32, error) {
	return c.getBalance()
}

// Method to get Queue ID from the API.
func (c Capmonster) getID(data *AdditionalData) (int, error) {
	// Get Payload
	payload, err := c.createPayload(data)
	if err != nil {
		return 0, err
	}

	// Make request to get answer
	for i := 0; i < 100; i++ {
		resp, err := http.Post("https://api.capmonster.cloud/createTask", "application/json", bytes.NewBuffer([]byte(payload)))
		if err != nil {
			time.Sleep(3 * time.Second)
			continue
		}
		body, _ := io.ReadAll(resp.Body)
		response := &capmonsterIDResponse{}
		resp.Body.Close()
		json.Unmarshal(body, response)

		// Parse the response
		if response.ErrorID != 0 { // Means there was an error
			return 0, errCodeToError(response.ErrorCode)
		}
		return response.TaskID, nil
	}
	return 0, ErrMaxAttempts
}

// This method gets the captcha token from the Capmonster API
func (c Capmonster) getCaptchaAnswer(ctx context.Context, additional ...*AdditionalData) (*CaptchaAnswer, error) {
	var data *AdditionalData = nil
	if len(additional) > 0 {
		data = additional[0]
	}

	// Get Queue ID
	queueID, err := c.getID(data)
	if err != nil {
		return nil, err
	}

	// Get Captcha Answer
	payload, _ := json.Marshal(capmonsterCapAnswerPayload{
		ClientKey: c.config.Api_key,
		TaskID:    queueID,
	})
	response := &capmonsterTokenResponse{}
	for i := 0; i < 100; i++ {
		req, _ := http.NewRequestWithContext(ctx, "POST", "https://api.capmonster.cloud/getTaskResult", bytes.NewBufferString(string(payload)))
		req.Header.Add("Content-Type", "application/json")
		resp, err := makeRequest(req)
		if err != nil {
			if errors.Is(err, context.DeadlineExceeded) {
				return nil, fmt.Errorf("getCaptchaAnswer error: %w", err)
			}
			time.Sleep(3 * time.Second)
			continue
		}

		// Parse Response
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		json.Unmarshal(body, response)

		// Check for any errors
		if response.ErrorID != 0 { // means there was an error
			return nil, errCodeToError(response.ErrorCode)
		}

		// Check if captcha is ready
		if response.Status == "processing" {
			time.Sleep(3 * time.Second)
			continue
		}

		var solution string
		var ua string = response.Solution.UserAgent
		switch c.config.CaptchaType {
		case V2Captcha, V3Captcha:
			solution = response.Solution.GRecaptchaResponse
		case ImageCaptcha:
			solution = response.Solution.Text
		case CFTurnstile:
			solution = response.Solution.Token
		}

		return newCaptchaAnswer(
			queueID,
			solution,
			c.config.Api_key,
			c.config.CaptchaType,
			AnticaptchaSite,
			ua,
		), nil
	}
	return nil, ErrMaxAttempts
}

// getBalance() returns the balance on the API key
func (c Capmonster) getBalance() (float32, error) {
	// Attempt to get the balance from the API
	// Max attempts is 5
	payload := fmt.Sprintf(`{"clientKey": "%v"}`, c.config.Api_key)
	response := &capmonsterBalanceResponse{}
	for i := 0; i < 5; i++ {
		resp, err := http.Post("https://api.capmonster.cloud/getBalance", "application/json", bytes.NewBuffer([]byte(payload)))
		if err != nil {
			time.Sleep(1 * time.Second)
			continue
		}

		// Parse Response
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		json.Unmarshal(body, response)
		if response.ErrorID != 0 {
			return 0, errCodeToError(response.ErrorCode)
		}
		return response.Balance, nil
	}
	return 0, ErrMaxAttempts
}

/*
createPayload returns the payloads required to interact with the API.

Possible errors that can be returned:
1) ErrIncorrectCapType
*/
func (c Capmonster) createPayload(data *AdditionalData) (string, error) {
	// Define the payload we are going to send to the API
	payload := capmonsterIDPayload{
		ClientKey: c.config.Api_key,
		Task: struct {
			WebsiteURL    string      "json:\"websiteURL\""
			WebsiteKey    string      "json:\"websiteKey\""
			Type          captchaType "json:\"type\""
			IsInvisible   bool        "json:\"isInvisible,omitempty\""
			MinScore      float32     "json:\"minScore,omitempty\""
			PageAction    string      "json:\"pageAction,omitempty\""
			Body          string      "json:\"body,omitempty\""
			ProxyType     string      "json:\"proxyType,omitempty\""
			ProxyAddress  string      "json:\"proxyAddress,omitempty\""
			ProxyPort     int         "json:\"proxyPort,omitempty\""
			ProxyLogin    string      "json:\"proxyLogin,omitempty\""
			ProxyPassword string      "json:\"proxyPassword,omitempty\""
			UserAgent     string      "json:\"userAgent,omitempty\""
		}{
			WebsiteURL: c.config.CaptchaURL,
			WebsiteKey: c.config.Sitekey,
			Type:       c.config.CaptchaType,
		},
	}

	// Add any other keys to the payload
	switch c.config.CaptchaType {
	case ImageCaptcha:
		if data == nil {
			return "", ErrAddionalDataMissing
		}
		payload.Task.Type = "ImageToTextTask"
		payload.Task.Body = data.B64Img
	case V2Captcha:
		payload.Task.Type = "NoCaptchaTaskProxyless"

		// Check for proxy data
		if data != nil && data.Proxy != nil {
			payload.Task.Type = "NoCaptchaTask"
			if data.ProxyType == "" {
				data.ProxyType = "http"
			}
			payload.Task.ProxyType = data.ProxyType
			payload.Task.ProxyAddress = data.Proxy.Ip
			portInt, err := strconv.Atoi(data.Proxy.Port)
			if err != nil {
				return "", errors.New("error converting proxy port to int")
			}
			payload.Task.ProxyPort = portInt
			if data.Proxy.IsUserAuth() {
				payload.Task.ProxyLogin = data.Proxy.User
				payload.Task.ProxyPassword = data.Proxy.Password
			}
		}

		if c.config.IsInvisibleCaptcha {
			payload.Task.IsInvisible = c.config.IsInvisibleCaptcha
		}
	case V3Captcha:
		payload.Task.Type = "RecaptchaV3TaskProxyless"
		payload.Task.MinScore = c.config.MinScore
		payload.Task.PageAction = c.config.Action
	case HCaptcha:
		payload.Task.Type = "HCaptchaTaskProxyless"
	case CFTurnstile:
		payload.Task.Type = "TurnstileTaskProxyless"
	default:
		return "", ErrIncorrectCapType
	}

	// Check for addtional data
	if data != nil && data.UserAgent != "" {
		payload.Task.UserAgent = data.UserAgent
	}

	encoded, _ := json.Marshal(payload)
	return string(encoded), nil
}
